Game flow rework (pt 10)

rr- opened the pull request:
#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description

- Fixes #2404 
- Shortens `GAME_FLOW_` type prefix to `GF_`
- Splits `GF_SEQUENCE_CONTEXT` from  `GAME_FLOW_LEVEL_TYPE`
- Streamlines accesses to the game flow levels by introducing better accessors in TRX
- Removes `g_CurrentLevel` in favor of `Game_GetCurrentLevel` (last played game level) and `GF_GetCurrentLevel` (any level, including cutscens and demos)
- Introduces level tables to reduce duplication in the reader and similar places
- Merges `GF_StorySoFar` to `GF_InterpretSequence` in TR1
- Merges almost everything, short of the sequence handlers and some very game-specific bits, via TRX

`reader.c` is a bit `#if`-ridden. I hope that we soon reach feature parity in both games in this field, so that this split can go away.

We still have the big difference in how we apply save game and resume info, eg. there's still the `load_level` event type in TR1 and the `Level_Initialise(…, GF_SEQUENCE_CONTEXT seq_ctx)` extra argument in TR2. This is something I'm leaving for the next PR as streamlining this is by far the trickiest part.

rr- committed:
tr1/game-flow: put sequence arrays earlier

rr- committed:
game-flow: introduce GAME_FLOW_SEQUENCE_CONTEXT

rr- committed:
tr1/game-flow: merge story sequence interpreter

rr- committed:
tr1/game-flow: use event handler mapping

rr- committed:
tr2/game-flow: use event handler mapping

rr- committed:
tr1/game-flow: tidy the api

rr- committed:
game-flow: move vars to libtrx

rr- committed:
game-flow: move GF_Shutdown to libtrx

rr- committed:
game-flow: streamline level access

rr- committed:
enum-map: add missing game flow defines

rr- committed:
game-flow: shorten GAME_FLOW_ prefix to just GF_

rr- committed:
game-flow: merge GF_START_GYM to GF_START_GAME

rr- committed:
game-flow: merge reader via libtrx

rr- committed:
game-flow: move common bits to libtrx

rr- committed:
tr1/game-flow: merge Game_Start_Legacy to GFS_LOAD_LEVEL

rr- commented:
Found a crash related to guns and resume info. Will address shortly

Edit: this is also visible in 4.7:

1. `/load 1`
2. `/play 8`
3. Repeat

rr- committed:
tr1/game-flow: fix crash with /load and /play

Resolves #2411.

aredfan commented:
Just one issue to report for TR1X - the fade-ins and fade-outs are now faster during credits.

rr- commented:
> Just one issue to report for TR1X - the fade-ins and fade-outs are now faster during credits.

Fixed. Also fixed the Dragon's Lair FMV issue Lahm reported on Discord.

Potential areas where to look for breakage:
- normal gameplay succession: fmvs, cutscenes, levels, stats, global stats
- select level feature
- restart level feature
- bonus levels
- passport interactions
- enemy item drops
- saving and loading (extensive – with focus on items carry over)
- saving and loading in gym via /save
- `/play name`, `/play id` (separately)
- `/demo id`
- `/title`
- `/load`, `/save` as well as passport (different routes)
- story so far (fmvs from the start of the saved level should play, fmvs coming after the saved level shouldn't play yet)

aredfan commented:
> Fixed. Also fixed the Dragon's Lair FMV issue Lahm reported on Discord.

Thank you. 👍 

Just a quick mention that the TRUB credits fade-ins/outs are fast as well.

rr- committed:
tr1/savegame: fix death counter support

Resolves #2412.

rr- commented:
@aredfan fixed. Also, fixed #2412 hopefully, though existing saves will not work (as in they won't show the death counter at all, as the information was not valid to begin with).

aredfan commented:
> @aredfan fixed. Also, fixed #2412 hopefully, though existing saves will not work (as in they won't show the death counter at all, as the information was not valid to begin with).

Thank you again. I can confirm the death counter is now fixed when following the steps laid out in your issue.

I also came across 2 new issues:

- [x] new_game_plus_unlock doesn't change from false to true post-credits. 

1. `/set new 0`
2. `/play 15`
3. use the level skip cheat and go through the credits.

- [x] Accessing Select Level via Load Game on the main menu crashes the game.

rr- commented:
Thanks @aredfan addressed both issues.
